File: solutions/ruby/sum-of-multiples/1/sum_of_multiples.rb
class SumOfMultiples
    def initialize(*base_values)
        @values = base_values
    end

    def to(level)
        return 0 if @values.all? { | value | level < value }

        multiples = Set.new
        @values.filter { | value | value > 0 }.each do
            | value |
            multiples.merge(find_multiples_for(value, level))
        end

        multiples.sum
    end

    private
    def find_multiples_for(start, limit)
        multiples = []
        multiple = start
        i = 2
        while multiple < limit
            multiples << multiple
            multiple = start * i
            i += 1
        end

        # puts "Multiples for #{start} to #{limit} = #{multiples}"
        multiples

    end
end

File: solutions/elixir/sum-of-multiples/1/lib/sum_of_multiples.ex
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    factors
    |> Enum.reduce([], fn number, acc -> acc ++ factors_to(number, limit) end)
    |> MapSet.new()
    |> Enum.sum()
  end

  defp factors_to(0, _), do: [0]
  defp factors_to(start, limit), do: Enum.to_list(start..limit-1//start)
end
